File: backend/services/chroma_service.py
import chromadb
from chromadb.utils import embedding_functions
from config.config import settings
from typing import List, Dict, Any, Optional
import asyncio
import uuid
from chromadb.api.types import QueryResult

class ChromaService:
    def __init__(self, host: str = "http://localhost:8000"):
        if not settings.CHROMA_URL:
            raise ValueError("CHROMA_URL is not set in the environment variables.")
            
        # CHROMA_URL might be like "http://localhost:8000"
        # The client needs host and port separately.
        try:
            url_parts = settings.CHROMA_URL.split(':')
            host = url_parts[1].replace('//', '')
            port = int(url_parts[2])
            
            self.client = chromadb.HttpClient(host=host, port=port)
            
            # Using a default embedding function for now, but the actual embeddings
            # will be provided by TitanService.
            self.default_ef = embedding_functions.DefaultEmbeddingFunction()
            print("ChromaDB client initialized.")

        except (IndexError, ValueError) as e:
            raise ValueError(f"Invalid CHROMA_URL format: {settings.CHROMA_URL}. Expected format: http://hostname:port. Error: {e}")
        except Exception as e:
            raise RuntimeError(f"Failed to initialize ChromaDB client: {e}")

    async def _get_collection(self, name: str):
        if not self.client:
            raise ConnectionError("ChromaDB client is not initialized.")
        try:
            # This is a synchronous call, run it in a thread
            return await asyncio.to_thread(self.client.get_or_create_collection, name=name)
        except Exception as e:
            print(f"Error getting or creating collection '{name}': {e}")
            raise

    async def list_collections(self):
        if not self.client:
            raise ConnectionError("ChromaDB client is not initialized.")
        return await asyncio.to_thread(self.client.list_collections)

    async def delete_collection(self, collection_name: str):
        if not self.client:
            raise ConnectionError("ChromaDB client is not initialized.")
        try:
            await asyncio.to_thread(self.client.delete_collection, name=collection_name)
            print(f"Collection {collection_name} deleted successfully.")
        except Exception as e:
            # ChromaDB's http client might raise various exceptions.
            # It's safer to catch a broad exception and log it.
            print(f"An error occurred while deleting collection {collection_name}: {e}")
            # Depending on requirements, you might want to re-raise or handle differently
            # For now, we print and absorb to prevent crashing the caller.
            # Re-raising a more specific, handled error might be better.
            raise ValueError(f"Failed to delete collection {collection_name}")

    async def query_collection(self, collection_name: str, query_embeddings: Any, n_results: int = 5) -> Optional[QueryResult]:
        """
        Queries a collection with the given embeddings.
        Runs the sync query in a separate thread.
        """
        try:
            collection = await self._get_collection(collection_name)
            results = await asyncio.to_thread(
                collection.query,
                query_embeddings=query_embeddings,
                n_results=n_results,
                include=["metadatas", "documents", "distances"]
            )
            return results
        except Exception as e:
            print(f"Error querying collection '{collection_name}': {e}")
            return None # Return None on error to be handled by the caller

    async def add_to_collection(self, collection_name: str, documents: List[str], embeddings: Any, metadatas: Any, ids: List[str]):
        if not documents:
            print("No documents to add. Skipping.")
            return
        
        collection = await self._get_collection(collection_name)
        
        try:
            await asyncio.to_thread(
                collection.add,
                embeddings=embeddings,
                documents=documents,
                metadatas=metadatas,
                ids=ids
            )
        except Exception as e:
            print(f"Error adding to collection '{collection_name}': {e}")
            raise

    async def add_documents(self, collection_name: str, documents_with_embeddings: List[Dict[str, Any]]):
        if not documents_with_embeddings:
            print("No documents to process. Skipping.")
            return
            
        collection = await self._get_collection(name=collection_name)

        # Unpack the list of document dictionaries into separate lists
        docs = [item['document'] for item in documents_with_embeddings]
        metadatas = [item['metadata'] for item in documents_with_embeddings]
        embeddings = [item['embedding'] for item in documents_with_embeddings]
        ids = [item['id'] for item in documents_with_embeddings]

        if not docs:
            print("Empty document list after unpacking. Skipping add to collection.")
            return
            
        try:
            await asyncio.to_thread(
                collection.add,
                ids=ids,
                embeddings=embeddings,
                documents=docs,
                metadatas=metadatas
            )
            print(f"Successfully added {len(docs)} documents to '{collection_name}'.")
        except Exception as e:
            print(f"Error during batch add to collection '{collection_name}': {e}")
            raise

    async def get_documents(self, collection_name: str, limit: int = 100, offset: int = 0):
        collection = await self._get_collection(collection_name)
        if not collection:
            return {"documents": [], "total": 0}

        total_count = await asyncio.to_thread(collection.count)
        
        results = await asyncio.to_thread(collection.get, limit=limit, offset=offset, include=["metadatas", "documents"])
        
        # The result 'ids' corresponds to the documents and metadatas
        # We'll zip them together to return a list of document objects
        docs = []
        if results and results.get('ids'):
             for i, doc_id in enumerate(results['ids']):
                doc_data = {
                    "id": doc_id,
                    "document": results['documents'][i] if results['documents'] and i < len(results['documents']) else None,
                    "metadata": results['metadatas'][i] if results['metadatas'] and i < len(results['metadatas']) else None
                }
                docs.append(doc_data)

        return {"documents": docs, "total": total_count}

    async def delete_documents(self, collection_name: str, document_ids: List[str]):
        if not document_ids:
            return
        collection = await self._get_collection(collection_name)
        if not collection:
            raise ValueError(f"Collection '{collection_name}' not found.")
            
        await asyncio.to_thread(collection.delete, ids=document_ids)

    async def delete_documents_by_source(self, collection_name: str, source_name: str):
        """Deletes all documents from a collection that match a given source name."""
        collection = await self._get_collection(collection_name)
        if not collection:
            raise ValueError(f"Collection '{collection_name}' not found.")

        # Find documents with the specified source metadata
        results = await asyncio.to_thread(
            collection.get,
            where={"source": source_name},
            include=[] # We only need the IDs
        )
        
        doc_ids_to_delete = results.get('ids')

        if not doc_ids_to_delete:
            print(f"No documents found with source '{source_name}' in collection '{collection_name}'.")
            return

        print(f"Found {len(doc_ids_to_delete)} documents from source '{source_name}' to delete.")
        await self.delete_documents(collection_name, doc_ids_to_delete)
        print(f"Successfully deleted documents from source '{source_name}'.")

# Instantiate the service
chroma_service = ChromaService() 

File: backend/src/services/chromaService.ts
import { ChromaClient, Collection, EmbeddingFunction } from 'chromadb';
import { realEmbeddingService } from '../core/realEmbeddingService';

// Real embedding function using AWS Bedrock
class RealEmbeddingFunction implements EmbeddingFunction {
  constructor() {}

  async generate(texts: string[]): Promise<number[][]> {
    try {
      console.log(`🔧 Generating real embeddings for ${texts.length} texts`);
      return await realEmbeddingService.embedBatch(texts);
    } catch (error) {
      console.error('❌ Failed to generate embeddings:', error);
      // Fallback to zero vectors only if absolutely necessary
      console.warn('⚠️ Using zero vectors as fallback');
      return texts.map(() => new Array(1536).fill(0));
    }
  }
}

export class ChromaService {
  private client: ChromaClient;
  private embeddingFunction: RealEmbeddingFunction;

  constructor() {
    const url = process.env.CHROMA_URL || 'http://localhost:8000';
    this.client = new ChromaClient({ path: url });
    this.embeddingFunction = new RealEmbeddingFunction();
    console.log('🚀 ChromaService initialized with real embeddings');
  }

  async getOrCreateCollection(name: string): Promise<Collection> {
    try {
      return await this.client.getOrCreateCollection({ 
        name,
        embeddingFunction: this.embeddingFunction
      });
    } catch (e) {
      console.error(`[ChromaService] Error getOrCreateCollection:`, e);
      throw e;
    }
  }

  async listCollections() {
    try {
      return await this.client.listCollections();
    } catch (e) {
      console.error(`[ChromaService] Error listCollections:`, e);
      throw e;
    }
  }

  async deleteCollection(collectionName: string) {
    try {
      return await this.client.deleteCollection({ name: collectionName });
    } catch (e) {
      console.error(`[ChromaService] Error deleteCollection:`, e);
      throw e;
    }
  }

  async addToCollection(collectionName: string, documents: string[], embeddings: number[][], metadatas: any[], ids: string[]) {
    if (!documents || documents.length === 0) {
      console.log('[ChromaService] No documents to add. Skipping.');
      return;
    }
    try {
      const collection = await this.getOrCreateCollection(collectionName);
      return await collection.add({
        ids,
        embeddings,
        documents,
        metadatas,
      });
    } catch (e) {
      console.error(`[ChromaService] Error addToCollection:`, e);
      throw e;
    }
  }

  async addDocuments(
    collectionName: string, 
    documentsWithEmbeddings: Array<{ document: string; embedding: number[]; metadata: any; id: string }>,
    onProgress?: (completed: number, total: number) => void
  ) {
    if (!documentsWithEmbeddings || documentsWithEmbeddings.length === 0) {
      console.log('[ChromaService] No documents to process. Skipping.');
      return;
    }

    console.log(`[ChromaService] Starting to add ${documentsWithEmbeddings.length} documents to '${collectionName}'`);
    
    // Dedup by id
    const seen = new Set();
    const docs = [];
    const metadatas = [];
    const embeddings = [];
    const ids = [];
    
    for (const item of documentsWithEmbeddings) {
      if (seen.has(item.id)) continue;
      seen.add(item.id);
      docs.push(item.document);
      metadatas.push(item.metadata);
      embeddings.push(item.embedding);
      ids.push(item.id);
    }
    
    if (docs.length === 0) {
      console.log('[ChromaService] All documents are duplicates. Skipping.');
      return;
    }

    try {
      // สำหรับข้อมูลจำนวนมาก ให้แบ่งเป็น batch เพื่อป้องกัน memory overflow
      if (docs.length > 100) {
        await this.addDocumentsInBatches(collectionName, {
          ids,
          embeddings,
          documents: docs,
          metadatas,
        }, onProgress);
      } else {
        const collection = await this.getOrCreateCollection(collectionName);
        await collection.add({
          ids,
          embeddings,
          documents: docs,
          metadatas,
        });
        onProgress?.(docs.length, docs.length);
      }
      
      console.log(`[ChromaService] Successfully added ${docs.length} documents to '${collectionName}'.`);
    } catch (e) {
      console.error(`[ChromaService] Error addDocuments:`, e);
      throw e;
    }
  }

  private async addDocumentsInBatches(
    collectionName: string,
    data: {
      ids: string[];
      embeddings: number[][];
      documents: string[];
      metadatas: any[];
    },
    onProgress?: (completed: number, total: number) => void
  ) {
    const BATCH_SIZE = 50; // จำนวน documents ต่อ batch
    const { ids, embeddings, documents, metadatas } = data;
    const total = documents.length;
    let completed = 0;

    console.log(`[ChromaService] Adding ${total} documents in batches of ${BATCH_SIZE}`);
    
    const collection = await this.getOrCreateCollection(collectionName);
    
    for (let i = 0; i < total; i += BATCH_SIZE) {
      const endIndex = Math.min(i + BATCH_SIZE, total);
      const batchIds = ids.slice(i, endIndex);
      const batchEmbeddings = embeddings.slice(i, endIndex);
      const batchDocuments = documents.slice(i, endIndex);
      const batchMetadatas = metadatas.slice(i, endIndex);
      
      try {
        console.log(`[ChromaService] Adding batch ${Math.floor(i / BATCH_SIZE) + 1}/${Math.ceil(total / BATCH_SIZE)}`);
        
        await collection.add({
          ids: batchIds,
          embeddings: batchEmbeddings,
          documents: batchDocuments,
          metadatas: batchMetadatas,
        });
        
        completed += batchIds.length;
        onProgress?.(completed, total);
        
        // เล็กน้อยหน่วงเวลาเพื่อป้องกัน overwhelm ChromaDB
        if (i + BATCH_SIZE < total) {
          await new Promise(resolve => setTimeout(resolve, 100));
        }
        
      } catch (error) {
        console.error(`[ChromaService] Error in batch ${Math.floor(i / BATCH_SIZE) + 1}:`, error);
        throw error;
      }
    }
    
    console.log(`[ChromaService] Successfully added all ${total} documents in batches`);
  }

  async queryCollection(collectionName: string, queryEmbeddings: number[][], nResults: number = 5) {
    try {
      const collection = await this.getOrCreateCollection(collectionName);
      const result = await collection.query({
        queryEmbeddings,
        nResults,
        include: ["metadatas", "documents", "distances"],
      });
      
      // ตรวจสอบและปรับปรุง response format
      if (result && result.documents) {
        console.log(`[ChromaService] Query successful for '${collectionName}': found ${result.documents.flat().length} documents`);
        return result;
      } else {
        console.log(`[ChromaService] No results found for '${collectionName}'`);
        return {
          documents: [],
          metadatas: [],
          distances: []
        };
      }
    } catch (e) {
      console.error(`[ChromaService] Error queryCollection for '${collectionName}':`, e);
      return {
        documents: [],
        metadatas: [],
        distances: []
      };
    }
  }

  async getDocuments(collectionName: string, limit: number = 100, offset: number = 0) {
    try {
      const collection = await this.getOrCreateCollection(collectionName);
      if (!collection) {
        console.log(`[ChromaService] Collection '${collectionName}' not found`);
        return { documents: [], total: 0 };
      }

      const all = await collection.get();
      const totalCount = all.ids?.length || 0;
      
      const docs = [];
      if (all && all.ids) {
        for (let i = offset; i < Math.min(offset + limit, all.ids.length); i++) {
          docs.push({
            id: all.ids[i],
            document: all.documents?.[i],
            metadata: all.metadatas?.[i],
          });
        }
      }
      
      // Sample documents log
      if (docs.length > 0) {
        console.log(`[ChromaService] Sample documents in '${collectionName}':`);
        for (let i = 0; i < Math.min(2, docs.length); i++) {
          console.log(`  ${i + 1}. ${docs[i].document?.slice(0, 100)}...`);
        }
      }
      
      return { documents: docs, total: totalCount };
    } catch (e) {
      console.error(`[ChromaService] Error getDocuments:`, e);
      return { documents: [], total: 0 };
    }
  }

  async deleteDocuments(collectionName: string, documentIds: string[]) {
    if (!documentIds || documentIds.length === 0) return;
    try {
      const collection = await this.getOrCreateCollection(collectionName);
      return await collection.delete({ ids: documentIds });
    } catch (e) {
      console.error(`[ChromaService] Error deleteDocuments:`, e);
      throw e;
    }
  }

  // Check if a document exists by id in a collection
  async documentExists(collectionName: string, id: string): Promise<boolean> {
    try {
      const collection = await this.getOrCreateCollection(collectionName);
      const res = await collection.get({ ids: [id] });
      return Array.isArray(res?.ids) && res.ids.length > 0;
    } catch (e) {
      console.error(`[ChromaService] Error documentExists:`, e);
      return false;
    }
  }

  async deleteDocumentsBySource(collectionName: string, sourceName: string) {
    try {
      const collection = await this.getOrCreateCollection(collectionName);
      const results = await collection.get({ where: { source: sourceName } });
      const docIdsToDelete = results.ids;
      if (!docIdsToDelete || docIdsToDelete.length === 0) {
        console.log(`[ChromaService] No documents found with source '${sourceName}' in collection '${collectionName}'.`);
        return;
      }
      await this.deleteDocuments(collectionName, docIdsToDelete);
      console.log(`[ChromaService] Successfully deleted documents from source '${sourceName}'.`);
    } catch (e) {
      console.error(`[ChromaService] Error deleteDocumentsBySource:`, e);
      throw e;
    }
  }

  // ดึงข้อความทั้งหมดจาก collection (vectorstore)
  async getAllFromCollection(collectionName: string): Promise<Array<{ document: string | null; metadata: any }>> {
    try {
      const collection = await this.getOrCreateCollection(collectionName);
      const all = await collection.get();
      const docs = [];
      
      if (all && all.ids && all.ids.length > 0) {
        for (let i = 0; i < all.ids.length; i++) {
          docs.push({
            id: all.ids[i],
            document: all.documents?.[i] ?? null,
            metadata: all.metadatas?.[i] || {},
          });
        }
        console.log(`[ChromaService] Retrieved ${docs.length} documents from '${collectionName}'`);
      } else {
        console.log(`[ChromaService] No documents found in '${collectionName}'`);
      }
      
      return docs;
    } catch (e) {
      console.error(`[ChromaService] Error getAllFromCollection for '${collectionName}':`, e);
      return [];
    }
  }

  // Placeholder for LangChain integration (option)
  getVectorStore(collectionName: string) {
    // Not implemented in JS version, but can be extended for LangChain/JS
    console.log(`[ChromaService] getVectorStore is not implemented in this version.`);
    return null;
  }
}

export const chromaService = new ChromaService(); 